### ISSUE TITLE
[popup] "Interest"-based triggering for popovers

### ISSUE THREAD

**Issue by mfreed7** [openui/open-ui#526]
The explainer currently [says this](https://github.com/openui/open-ui/blob/main/research/src/pages/popup/popup.research.explainer.mdx#classes-of-ui---dismiss-behavior-and-interactions) about the behavior for `popup=hint` (which I'll call "tooltips" for clarity):

> Hint/Tooltip (popup=hint)
> - When opened, force-closes only other hints, but leaves all other popup types open.
> - Dismisses on [close signal](https://wicg.github.io/close-watcher/#close-signal), click outside, when no longer hovered (after a timeout), or when the anchor element loses focus.

For the light dismiss triggers, everything except `when no longer hovered` is well defined. However, since we haven't discussed a declarative way to **trigger** tooltips, it isn't clear how "no longer hovered" should be defined. And even if there was a declarative trigger, what happens for tooltips that are programmatically triggered via `tooltip.showPopup()`? In order to answer questions about light dismiss for tooltips, I think we need to figure out how declarative invocation works.

Here's a quick strawman off-the-cuff idea:
```html
<img src=questionmark.png hoverhint=foo>
<div popup=hint id=foo>Tooltip</div>
<style>
img {
  hint-trigger-delay=500ms;
}
[popup=hint] {
  hint-hide-delay=1000ms;
}
</style>
```

In other words, there is a new content attribute, `hoverhint`, which points to an element with `popup=hint`. When the `<img>` element (in this example, but could be any element type) is hovered, after a delay set by the `hint-trigger-delay` CSS property, the tooltip is shown. When that tooltip (or the triggering `<img>` element) has no longer been hovered for a time set by `hint-hide-delay`, it is hidden.

Again, this is a total strawman proposal. Suggestions appreciated.

**Comment by scottaohara** [comment]
i think i'd just expect a `:hover` and `:focus` for the selector to tie it to those actions.  As otherwise i'd read that as expecting it to show by default after the 500ms delay.

e.g.,
```html
*:hover,
*:focus {
  hint-trigger-delay=500ms;
}
```

What I don't readily have answer for is how non-interactive elements will allow users who aren't using a mouse to access their tooltips (e.g., your image example) or how these tooltips will be accessible on touch screens.

What might be useful is if caret navigation had it's own pseudo-class similar to `:hover`, or tapped into`:hover`, as that could be a means to trigger a hint on an otherwise non-focusable element.  

I only have partial thoughts on what could be done for touch screen access.... I'll have to think more on that / we probably want to bring this up to a larger group to say "hey, we're talking about tooltips.... they've been horribly accessible on non-interactive elements for the last 24+ years... how we going to fix that?"

**Comment by mfreed7** [comment]
> ```
> *:hover,
> *:focus {
>   hint-trigger-delay=500ms;
> }
> ```
 
So I think that would be fine, since the `hint-trigger-delay` represents the time that the user needs to hover or focus the element before it triggers the tooltip. So it'll only matter if/when the element is hovered/focused. Having said that, it also shouldn't matter if it was selected like `* {hint-trigger-delay=500ms}` because it'll have no effect when the element isn't hovered/focused. Perhaps your point was just that we need to make sure to include `focus`, and I totally agree with that.

> What I don't readily have answer for is how non-interactive elements will allow users who aren't using a mouse to access their tooltips (e.g., your image example) or how these tooltips will be accessible on touch screens.

Totally agree with this. Suggestions appreciated.

**Comment by mfreed7** [comment]
Great discussion yesterday, with several good comments. One important one, which we didn't resolve on, but sounded like it had general agreement was:

- Instead of a new attribute (`hoverhint`), let's just re-use `togglepopup`, `showpopup`, and `hidepopup`.

I like the idea of re-using the existing attributes. There are a few questions we'd have to answer:
- When applied to non-button-like elements, that should only work when the target popup is a `popup=hint`. What happens otherwise, nothing?
- What is the meaning of `hidepopup=foo` when `foo` is a `popup=hint`? Does it immediately hide the popup if it's showing, rather than waiting for the delay to expire? I'm not sure what the use case is.
- What exactly does `togglepopup=foo` mean also? Hover it once and the tooltip shows, hover it again and it goes away? Feels odd.
- Given the above two, maybe only `showpopup` makes sense?

-----

There was also some discussion of a11y mappings for `popup=hint`. Let's discuss that on https://github.com/openui/open-ui/issues/329.

There was some discussion of keyboard support for triggering `popup=hint`, and @scottaohara nicely opened an issue to discuss that: https://github.com/openui/open-ui/issues/530.

Another point was made about naming: `hint` and `async` are confusing because they have to be mentally mapped to "tooltip" and "toast/notification/something". Perhaps we should pick better names. I've opened https://github.com/openui/open-ui/issues/532 to discuss that.

**Comment by aleventhal** [comment]
Is it settled that hint popups will be triggered on focus?

**Comment by mfreed7** [comment]
Based on the [resolution above](https://github.com/openui/open-ui/issues/526#issuecomment-1158012026) (not the [first one](https://github.com/openui/open-ui/issues/526#issuecomment-1157997881), which was inadvertently posted here), we've decided to go back to a unique name for the declarative triggering of hint popups.

I'd like to re-post the proposal from the OP:

```html
<img src=questionmark.png hoverhint=foo>   <!-- hoverhint means foo gets shown when this img is hovered -->
<div popup=hint id=foo>Tooltip</div>      <!-- hint pop up, but maybe could also be auto or async pop up? -->
<style>
img {
  hint-trigger-delay=500ms;  /* Controls how long the img needs to be hovered before triggering the pop up */
}
[popup=hint] {
  hint-hide-delay=1000ms; /* After the pop up/img are no longer hovered for this long, hide the pop up */
}
</style>
```

The code snippet above has some comments, which also incorporate part of the discussion we had last week. In particular, these behaviors might "just work" for any kind of pop up, not only `popup=hint` (or [whatever](https://github.com/openui/open-ui/issues/532)). I.e. `hoverhint` could point to any kind of pop up to trigger-on-hover. And if the CSS `hint-hide-delay` property was set on a `popup` element (any kind?), then de-hovering for that long triggers a hide.

Thoughts? Suggestions for improvement?

**Comment by VicGUTT** [comment]
In

```css
img {
  hint-trigger-delay=500ms;
}
[popup=hint] {
  hint-hide-delay=1000ms;
}
```

I would suggest parity with the terms used for activation/deactivation of popup (aka: show/hide).
So: `hint-show-delay` and `hint-hide-delay`. Plus using the opposite of "hide" here creates an intuitive parallel between the two words I believe.

> In particular, these behaviors might "just work" for any kind of pop up, not only `popup=hint` (or [whatever](https://github.com/openui/open-ui/issues/532)). I.e. `hoverhint` could point to any kind of pop up to trigger-on-hover.

Well in that case all "hint" references should be replaced by "popup":

```html
<img src=questionmark.png hoverpopup=foo>   <!-- hoverpopup means foo gets shown when this img is hovered -->
<div popup=async id=foo>Tooltip</div>      <!-- async pop up, but could also be auto or hint pop up -->
<style>
img {
  popup-show-delay: 500ms;  /* Controls how long the img needs to be hovered before triggering the pop up */
}
[popup=hint] {
  popup-hide-delay: 1000ms; /* After the pop up/img are no longer hovered for this long, hide the pop up */
}
</style>
```

---

My personal open question for the very specific CSS properties (`popup-(show|hide)-delay`):

Is it becoming common to have CSS properties pertain to specific HTML contexts/elements/attributes ?

Wouldn't the more CSS'y way to add a new pseudo-class ? Something like:

```css
img:popup-show {
  transition-delay: 500ms;
}
```

I know I saw some early specs (for anchor positioning maybe ?) also having some CSS properties expecting a very specific HTML context. So I'm just wondering.

**Comment by mfreed7** [comment]
> I would suggest parity with the terms used for activation/deactivation of popup (aka: show/hide). So: `hint-show-delay` and `hint-hide-delay`. Plus using the opposite of "hide" here creates an intuitive parallel between the two words I believe.

Thanks - I think I agree with this point. It'd be nice to have these be more symmetrical.


> > In particular, these behaviors might "just work" for any kind of pop up, not only `popup=hint` (or [whatever](https://github.com/openui/open-ui/issues/532)). I.e. `hoverhint` could point to any kind of pop up to trigger-on-hover.
> 
> Well in that case all "hint" references should be replaced by "popup":

100% agree with this one. In fact, as I've been prototyping this in Chromium, I used `hoverpopup` as the name, and made it work for all pop-up types. Taking this together with the above, and adding dashes for "pop-up", we're at:

- `pop-up-show-delay` - hover a **triggering element** for this long to show the pop-up
- `pop-up-hide-delay` - **de**-hover the **pop-up** or its triggers/anchors for this long to hide the pop-up.


> My personal open question for the very specific CSS properties (`popup-(show|hide)-delay`):
> 
> Is it becoming common to have CSS properties pertain to specific HTML contexts/elements/attributes ?

I agree that this does feel a bit weird: `pop-up-show-delay` only has an effect on elements that have the `hoverpopup` attribute. And `pop-up-hide-delay` only has an effect on elements that have the `popup` attribute. But I guess I don't see how that's terribly different from CSS properties that only have an effect when other CSS properties are also present, such as `transition-delay`. That only does anything when `transition-duration` != 0.

> Wouldn't the more CSS'y way to add a new pseudo-class ? Something like:
> 
> ```css
> img:popup-show {
>   transition-delay: 500ms;
> }
> ```

That's an interesting suggestion. The problem is that it's a bit muddled - what state/class does the `:popup-show` pseudo class match? And that would preclude using `transition-delay` on that element for actual transitions, right?

**Comment by VicGUTT** [comment]
Ah, sorry, only just now seeing your comment. I should look into configuring my Github notifications.

Yes, the "_dashyfied_" version of popup. So yep, all for `pop-up-show-delay` and `pop-up-hide-delay`.

> I agree that this does feel a bit weird [...] But I guess I don't see how that's terribly different from CSS properties that only have an effect when other CSS properties are also present, such as `transition-delay`. That only does anything when `transition-duration` != 0.

I don't see it being "**terribly** different" either but "different" nonetheless.

- First, the dependency is in relation with another CSS property and stays within the CSS only context.
- In this particular example they share a relation as prefixed by the same `transition` keyword. Also, they can be condensed into shorthand versions `transition: ...;`
- Here we at least have precedence (CSS properties depending on each other).

I guess my question was more in the lines of "is there precedence" or "is this a agreed upon / reliable strategy moving forward, where CSS and HTML become so implicitly inter-dependent".

> That's an interesting suggestion. The problem is that it's a bit muddled - what state/class does the :popup-show pseudo class match?

Yes, I agree. More thoughts would need to go into it. This conversation could probably be tied to the `:top-layer/:modal` discussions.

>  And that would preclude using transition-delay on that element for actual transitions, right?

Right, good catch. I'd say "possibly", but it'd definitely create some conflicts.
I guess in that case maybe the `pop-up-(show|hide)-delay` properties could both be moved to the element with the `popup="..."` attribute.

Which begs the question: in it's inactive state, would the `[popup=hint]` element have a default display of none?
If yes, the problem could be broaden and "generic'ifed" to include any display none'd element?

**Comment by mfreed7** [comment]
The meeting above had some great discussion. Several folks had good suggestions for this part of the API - please post them here, and we'll discuss at a future meeting.

**Comment by mfreed7** [comment]
Any comments on this issue? I have the proposed behavior implemented in Chromium, with some renaming suggestions from the discussion above. As it stands, this is how it works:

```html
<div popuphovertarget=foo>
<div popup id=foo>Pop-up content</div>

<style>
[popuphovertarget] { pop-up-show-delay=500ms; }
[popup] { pop-up-hide-delay=1s; }
</style>
```

New concepts:

- `popuphovertarget=foo` means that when the element with this attribute is hovered, the pop-up with `id=foo` should be shown. This attribute can be applied to any element type.
- The `pop-up-show-delay` CSS property controls how **long** you have to hover before the pop-up is shown. This property has no effect on elements that do not have a `popuphovertarget` attribute, because there is nothing to show in that case.
- The `pop-up-hide-delay` CSS property enables a feature that will **hide** a pop-up after it has not been hovered for a period of time (controlled by the value of this property). This is **independent** of the above two points, and can be used on any pop-up.

We've also discussed (in [this meeting](https://github.com/openui/open-ui/issues/526#issuecomment-1191815864)) that the spec/behavior likely needs to include some affordance for non-mouse users. For example, non-interactive controls that have the `popuphovertarget` attribute might need to be added to the keyboard tab navigation order (i.e. act like `tabindex=0`) so that keyboard users can access this content. That would also mean it needs to be triggered in that case by the element getting focus. Perhaps all of that only happens when using caret navigation? For touch users, it sounded like there was already a way (e.g. via long touch) to activate the `:hover` pseudo class, so that use case might be handled.

Generally, thoughts appreciated on the above. Having implemented the prototype, it does seem quite useful, as you can implement a fully animated tooltip that shows and hides after customizable delays, all with about 4 lines of HTML/CSS and no JS.

**Comment by mfreed7** [comment]
So the TL;DR of the meeting just now was that while this might be a "cool" addition to the pop-up API, it still has many things that need to be resolved, including:
 - What to do about the `role` for elements with `popuphovertarget`?
 - How to handle non-mouse user interaction?
 - Should we call this "hover"? Or maybe something else that better captures the keyboard/touch/other use cases?
 - How to spec the implementation of the behavior: during hover chain update, during `mouseover` update, etc.? What happens if the CSS properties are inside `:hover` or `:focus` selectors - when are styles updated?

At this point, based on the discussion, I'm going to consider this a "V2" type feature, not to be included in the first version of the Pop-up API. In the meantime, we can use this issue to brainstorm the answers to the above questions plus any others I left out.

**Comment by mfreed7** [comment]
Alright, Popover v1 shipped. Let's start talking about the `popover=hint` related things again. I think this issue is a good starting point.

**Comment by brechtDR** [comment]
There are far greater A11Y experts here. But from what I understood was: 

tab on focus tooltip containing element -> next tab should toggle the tooltip with a focus on the container of the tooltip itself (by setting tabindex to 0 in this case).

using shift+tab to return, it should not toggle the tooltip.

escape key should close the tooltip.

Interested to hear thoughts on this 🙂

[Created this little codepen demo with key navigation idea](https://codepen.io/utilitybend/pen/gOXWLRo
)

**Comment by mfreed7** [comment]
Per my meeting action item: @scottaohara @aleventhal @smhigley  call to action! Let's figure out the right way to do "interest-based triggering" of popovers. See [this comment](https://github.com/openui/open-ui/issues/526#issuecomment-1219845155) for a possibly-incomplete list of open issues.

**Comment by mfreed7** [comment]
> So the TL;DR of the meeting just now was that while this might be a "cool" addition to the pop-up API, it still has many things that need to be resolved, including:
> 
> * What to do about the `role` for elements with `popuphovertarget`?
> * How to handle non-mouse user interaction?
> * Should we call this "hover"? Or maybe something else that better captures the keyboard/touch/other use cases?
> * How to spec the implementation of the behavior: during hover chain update, during `mouseover` update, etc.? What happens if the CSS properties are inside `:hover` or `:focus` selectors - when are styles updated?
> 
> At this point, based on the discussion, I'm going to consider this a "V2" type feature, not to be included in the first version of the Pop-up API. In the meantime, we can use this issue to brainstorm the answers to the above questions plus any others I left out.

There's now an [explainer for `popover=hint` and hover triggering](https://open-ui.org/components/popover-hint.research.explainer/). I believe it addresses many of the questions above. Since hover triggering is only supported on buttons, the "role" question goes away. The explainer addresses non-mouse user interactions via keyboard focus or long-press. Naming ("hover") is still an open question, as it always is, but there's https://github.com/openui/open-ui/issues/767 for that specifically. And exactly how to spec it - let's cross that bridge when we start to spec it.

Given the above, I'm going to close this issue. Feel free to open fresh issues to discuss problems related to the [new explainer](https://open-ui.org/components/popover-hint.research.explainer/).
